pyro/contrib/funsor/infer/tracetmc_elbo.py

```diff
@@ -21,28 +22,25 @@ def _get_trace(self, *args, **kwargs):
         raise ValueError("shouldn't be here")
 
     def differentiable_loss(self, model, guide, *args, **kwargs):
-        with enum(first_available_dim=-self.max_plate_nesting-1):
+        with plate(size=self.num_particles) if self.num_particles > 1 else contextlib.ExitStack(), \
+                enum(first_available_dim=(-self.max_plate_nesting-1) if self.max_plate_nesting else None):
             guide_tr = trace(guide).get_trace(*args, **kwargs)
             model_tr = trace(replay(model, trace=guide_tr)).get_trace(*args, **kwargs)
 
-        log_factors, log_measures, measure_vars, plate_vars = [], [], frozenset(), frozenset()
-        for role, tr in zip(("model", "guide"), map(prune_subsample_sites, (model_tr, guide_tr))):
-            for name, node in tr.nodes.items():
-                if node["type"] != "sample":
-                    continue
-                # if a site is enumerated in the model, measure but no log_prob
-                if name in guide_tr.nodes or node['is_observed']:
-                    log_factors.append(
-                        node["funsor"]["log_prob"] if role == "model" else -node["funsor"]["log_prob"])
-                if node["funsor"].get("log_measure", None) is not None:
-                    log_measures.append(node["funsor"]["log_measure"])
-                    measure_vars |= frozenset(node["funsor"]["log_measure"].inputs)
-                plate_vars |= frozenset(f.name for f in node["cond_indep_stack"] if f.vectorized)
-                measure_vars |= frozenset(node["funsor"]["log_prob"].inputs)
+        model_terms = terms_from_trace(model_tr)
+        guide_terms = terms_from_trace(guide_tr)
```

Comment:
It's not obvious what the right way to handle scaling is in `TraceTMC_ELBO` - perhaps subsampling-induced rescaling should be distinguished from `handlers.scale` applications?.  Note that the original implementation (and its tests) punts on this as well, so it's probably out of scope for this PR.